### PR TITLE
feat: Feature/azure service principal 

### PR DESCRIPTION
**Pull request by FrsECM**
### PR fulfills these requirements
- [x] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
- [ ] Product design
- [x] Backend (Database)
- [x] Backend (API)
- [ ] Frontend

### Describe the reason for change
For the moment, there is only an authentication through account key to Azure.
The problem is that account key give a lot of rights on the storage that may not be necessary in order to perform readonly operations.

#### What does this fix?
Now it is possible to have an azure integration that is based on a service principal.

#### What is the new behavior?
We have to create an app registration in Azure : 

We give this registration specific rights depending on what we want to do in labelstudio : 

In the UI, we create a new storage integration : 

Under the hood : 
image

#### What is the current behavior?
There is no impact on the previous behavior, it's just a new one.

#### What libraries were added/updated?
Azure Identity

#### Does this change affect performance?
Get a delegation key is quite slow, that's why there is : 
- 24h cache for delegation key
- 15mn cache for sas token

There is a second minor change : 
- When a "thumbnail" tag is present in the datarow, it will be used in the grid view
- When there is multiple layer in the image, the gridview will only render the first channel.
(Today there is a buggy behavior).

#### Does this change affect security?
It should improve the security for people who just need specific rights on the container.
**_For the moment, the "write" behavior have not been tested / developped on the container._**

#### What alternative approaches were there?
We can use an enterprise application and SSO in order to be linked to a user right and act on behalf of him.

#### What feature flags were used to cover this change?
None

### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [ ] No
- [ ] Not sure (briefly explain the situation below)

### What level of testing was included in the change?
- [ ] e2e
- [ ] integration
- [ ] unit

### Which logical domain(s) does this change affect?
- Authentication
- UI

### CODE REVIEW

File: deploy/docker-entrypoint.sh [hunk 1]
```diff
@@ -69,7 +69,7 @@ if [ "$1" = "nginx" ]; then
   # in this mode we're running in a separate container
   export APP_HOST=${APP_HOST:=app}
   exec_entrypoint "$ENTRYPOINT_PATH/nginx/"
-  exec nginx -c $OPT_DIR/nginx/nginx.conf -e /dev/stderr
```
Comment (FrsECM) [inline deploy/docker-entrypoint.sh:72]: I got an issue with it in my containerisation saying that -e option is no longer supported by nginx.
At the same time, it's already in the conf.

File: web/libs/datamanager/src/components/MainView/GridView/GridView.js [hunk 2]
```diff
@@ -25,15 +25,35 @@ const GridHeader = observer(({ row, selected }) => {
...
+     */
+    if (field_type=='Image' && thumbnail.length>0)
+    {
+      return null;
+    }
+    /**The value is an array...
+     * In this case, we take the first element of the array
+     */
+    if (Array.isArray(value)){
+      value = value[0]
```
Comment (FrsECM) [inline web/libs/datamanager/src/components/MainView/GridView/GridView.js]: Fix bug when multilayered images.

File: web/libs/datamanager/src/components/MainView/GridView/GridView.js [hunk 3]
```diff
@@ -25,15 +25,35 @@ const GridHeader = observer(({ row, selected }) => {
 
 const GridBody = observer(({ row, fields }) => {
   const dataFields = fields.filter((f) => f.parent?.alias === "data");
+  const thumbnail = dataFields.filter(field=>field.alias=="thumbnail")
```
Comment (makseq) [inline web/libs/datamanager/src/components/MainView/GridView/GridView.js]: Please, don't mix two different ideas in the same PR. Let's split it to another PR.

**Review by makseq — CHANGES_REQUESTED** [review]
Could you also add pytest for this storage? There should be examples in tests/ folder for aws/gcs storages.

File: web/libs/datamanager/src/components/MainView/GridView/GridView.js (hunk 3 above)
Comment (FrsECM) [reply web/libs/datamanager/src/components/MainView/GridView/GridView.js]: The fix is done.
I created : https://github.com/HumanSignal/label-studio/pull/5769

**Comment by FrsECM** [comment]
I fixed the write access this morning.
You can now create an export integration :



Without the right right in Azure, you can't use 


But if you add : 
image

It will work and save the annotation :

**Comment by FrsECM** [comment]
> Could you also add pytest for this storage? There should be examples in tests/ folder for aws/gcs storages.

Can you help me on this part ?
I'am not sure to understand how i'm suppose to create tests.
The test i did locally were done by the UI and on my organisation private azure storage. I'm not able to replicate this in a public env.

**Comment by makseq** [comment]
Sorry for the long delay, tests are located here:
https://github.com/HumanSignal/label-studio/blob/develop/label_studio/tests/io_storages.tavern.yml#L1168

**Comment by FrsECM** [comment]
> Sorry for the long delay, tests are located here: https://github.com/HumanSignal/label-studio/blob/develop/label_studio/tests/io_storages.tavern.yml#L1168

Thanks,

I updated the **_get_import_export_storage_types_** test case to match the new syntax.

I saw that other things for gcp and s3, but in my case i have a strong validation that the connection is valid and can be consume, then i can't use dummy client_id/client_secret/tenant_id.

**Comment by sajarin** [comment]
/jira create
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/9289869847)
> Jira issue [TRIAG-578](https://humansignal.atlassian.net/browse/TRIAG-578) is created

[TRIAG-578]: https://humansignal.atlassian.net/browse/TRIAG-578?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by robot-ci-heartex** [comment]
This PR is stale because it has been open 45 days with no activity. Remove `stale` label or comment or this will be closed in 10 days.

**Comment by robot-ci-heartex** [comment]
This PR was closed because it has been stalled for 10 days with no activity.
